### PR TITLE
Add heartbeat and cool down period to instant-seal 💗

### PR DESCRIPTION
**Pull request by jimmychu0807**
This PR is to address #6849, and has received immense support from @JoshOrndorff and @seunlanlege 

We add a new heartbeat stream that merge the transaction pool notification stream in together. It has three characteristics:

- you can use instant-seal as before (meaning no heartbeat at all) if you pass a `None` in the heartbeat option in `run_instant_seal` function.

- There is a `max_blocktime` param in the heartbeat option. If no transactions come within a `max_blocktime` period, a new block will be generated. Default set to 30 sec.

- There is a `min_blocktime` param in the heartbeat option. If multiple transactions come within a `min_blocktime` period since the last block is generated, it will wait for the `min_blocktime` to be passed since the last block import and then seal a new block. This is to prevent block being sealed too frequently.

Before merge:
- [x] Add documentation
- [x] Remove code from node template

### CODE REVIEW

File: client/consensus/manual-seal/src/heartbeat_stream.rs [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+	fn default() -> Self {
+		Self {
+			timeout: DEFAULT_TIMEOUT,
+			min_blocktime: DEFAULT_MIN_BLOCKTIME,
+			finalize: DEFAULT_FINALIZE,
+		}
+	}
+}
+
+pub struct HeartbeatStream<Hash> {
```
Comment (JoshOrndorff) [inline client/consensus/manual-seal/src/heartbeat_stream.rs:39]: Needs Docs

File: client/consensus/manual-seal/src/heartbeat_stream.rs [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+// ---
+// Constant Definition
+
+const DEFAULT_TIMEOUT: u64 = 30;
+const DEFAULT_MIN_BLOCKTIME: u64 = 1;
+const DEFAULT_FINALIZE: bool = false;
+
+// ---
+pub struct HeartbeatOptions {
```
Comment (JoshOrndorff) [inline client/consensus/manual-seal/src/heartbeat_stream.rs:20]: Needs docs

File: client/consensus/manual-seal/src/heartbeat_stream.rs [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+	type Item = EngineCommand<Hash>;
+
+	fn poll_next(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
+		let mut hbs = self.get_mut();
+		match hbs.pool_stream.poll_next_unpin(cx) {
+			Poll::Ready(Some(ec)) => {
+				// Ensure we are not producing block too frequently, based on `min_blocktime` setting
+				if Instant::now() - hbs.instant < Duration::from_secs(hbs.opts.min_blocktime) {
+					return Poll::Pending;
+				}
```
Comment (JoshOrndorff) [inline client/consensus/manual-seal/src/heartbeat_stream.rs:71]: You've checked to make sure we don't author two blocks in short succession, but you haven't ensured that a second block will get authored in a reasonable amount of time. Imagine a node using instant seal. A user submits two transactions very close together. The first transaction causes a block to be authored. The second transaction does not cause a block to be authored because of this logic.

But that means the second transaction will sit in the pool until the next heartbeat block comes which will generally be a long time or possibly never if heartbeats are disabled. I think better behavior is that when a transaction arrives and it is too soon to author another block, a timer should be started so that the next block will be authored as soon the minimum block time runs out.

File: client/consensus/manual-seal/src/heartbeat_stream.rs [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+			timeout: DEFAULT_TIMEOUT,
+			min_blocktime: DEFAULT_MIN_BLOCKTIME,
+			finalize: DEFAULT_FINALIZE,
+		}
+	}
+}
+
+pub struct HeartbeatStream<Hash> {
+	pool_stream: Box<dyn Stream<Item = EngineCommand<Hash>> + Unpin + Send>,
+	delay: Delay,
```
Comment (JoshOrndorff) [inline client/consensus/manual-seal/src/heartbeat_stream.rs:41]: What is this delay field for? Is it different than the heartbeat period and min block period from the `HeartbeatOptions` struct?

File: client/consensus/manual-seal/src/heartbeat_stream.rs [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+						// We set `delay_future` here so those txs arrived after heartbeats doesn't have to wait
+						//   for `timeout`s to get processed, but only `min_blocktime`s.
+						hbs.delay_future = tokio::time::delay_for(Duration::from_secs(hbs.opts.min_blocktime));
+						return Poll::Pending;
+					}
+				}
+
+				// reset the timer and delay future
+				hbs.delay_future = tokio::time::delay_for(Duration::from_secs(hbs.opts.timeout));
+				hbs.last_heartbeat = Some(Instant::now());
```
Comment (seunlanlege) [inline client/consensus/manual-seal/src/heartbeat_stream.rs:84]: This throws away the engine command yeilded by pool_stream

File: client/consensus/manual-seal/src/heartbeat_stream.rs [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+				// We check if the delay for heartbeat has reached
+				if let Poll::Ready(_) = hbs.delay_future.poll_unpin(cx) {
+					// reset the timer and delay future
+					hbs.delay_future = tokio::time::delay_for(Duration::from_secs(hbs.opts.timeout));
+					hbs.last_heartbeat = Some(Instant::now());
+
+					return Poll::Ready(Some(EngineCommand::SealNewBlock {
+						create_empty: true, // heartbeat blocks are empty by definition
+						finalize: hbs.opts.finalize,
+						parent_hash: None, // QUESTION: no parent hash here? Is this block still conneected with the whole chain?
```
Comment (seunlanlege) [inline client/consensus/manual-seal/src/heartbeat_stream.rs:101]: parent_hash will be supplied by authorship task

File: client/consensus/manual-seal/src/heartbeat_stream.rs [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+				if let Some(last_heartbeat) = hbs.last_heartbeat {
+					// If the last heartbeat happened within min_blocktime time, we want to wait at least
+					//   until `min_blocktime` has passed.
+					if Instant::now() - last_heartbeat < Duration::from_secs(hbs.opts.min_blocktime) {
+						// We set `delay_future` here so those txs arrived after heartbeats doesn't have to wait
+						//   for `timeout`s to get processed, but only `min_blocktime`s.
+						hbs.delay_future = tokio::time::delay_for(Duration::from_secs(hbs.opts.min_blocktime));
+						return Poll::Pending;
+					}
+				}
```
Comment (seunlanlege) [inline client/consensus/manual-seal/src/heartbeat_stream.rs:80]: This throws away the engine command yeilded by pool_stream

File: client/consensus/manual-seal/src/lib.rs [hunk 8]
```diff
@@ -156,6 +170,9 @@ pub async fn run_instant_seal<B, CB, E, C, A, SC, T>(
 	pool: Arc<txpool::Pool<A>>,
 	select_chain: SC,
 	inherent_data_providers: InherentDataProviders,
+	finalize: bool,
+	heartbeat_opts: Option<HeartbeatOptions>,
+	_replies_sender: Option<mpsc::Sender<()>>,
```
Comment (seunlanlege) [inline client/consensus/manual-seal/src/lib.rs:175]: unused param

**Comment by jimmychu0807** [comment]
@seunlanlege I believe I will need your help from here 🙏

1. About your comment that `Some(ec)` will be thrown away, you are right. But how do I just push back the ec into the head of `pool_stream`? 

2. I have done my best on writing the tests. Please review the last two tests. I am particularly uncomfortable how the last test is written (the test on verifying min_blocktime).

If you want, feel free to commit on top of this PR. Thank you!

File: client/consensus/manual-seal/src/heartbeat_stream.rs (hunk 6 above)
Comment (JoshOrndorff) [reply client/consensus/manual-seal/src/heartbeat_stream.rs]: When you put `None` it uses the best known block.

**Review by jimmychu0807 — COMMENTED** [review]
@bkchr  @kianenigma Please let me know what you think of the PR. Thanks.

File: client/consensus/manual-seal/src/lib.rs [hunk 9]
```diff
@@ -19,7 +19,13 @@
 //! A manual sealing engine: the engine listens for rpc calls to seal blocks and create forks.
 //! This is suitable for a testing environment.
 
+#![allow(dead_code)]
```
Comment (JoshOrndorff) [inline client/consensus/manual-seal/src/lib.rs:22]: What is this for?

Comment (jimmychu0807) [reply client/consensus/manual-seal/src/lib.rs:22]: Because manual-seal is not used by default. If we don't have this line of code, compiler will warn us the code is there but not being used.

File: client/consensus/manual-seal/src/finalize_block.rs [hunk 10]
```diff
@@ -1,18 +1,20 @@
-// Copyright 2019 Parity Technologies (UK) Ltd.
```
Comment (kianenigma) [inline client/consensus/manual-seal/src/finalize_block.rs:1]: was there any purpose that we have license change in the PR? It seems the old code was using the old format, maybe it was for a reason?

File: client/consensus/manual-seal/src/heartbeat_stream.rs [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! HeartbeatStream type and HeartbeatOptions used in instant seal
```
Comment (kianenigma) [inline client/consensus/manual-seal/src/heartbeat_stream.rs:19]:
```suggestion
//! HeartbeatStream type and HeartbeatOptions used in instant seal

```

Comment (kianenigma) [reply client/consensus/manual-seal/src/heartbeat_stream.rs:19]: + this is a good spot to document how the whole heartbeat process is working.

File: client/consensus/manual-seal/src/heartbeat_stream.rs [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+		Self {
+			min_blocktime: DEFAULT_MIN_BLOCKTIME,
+			max_blocktime: DEFAULT_MAX_BLOCKTIME,
+			finalize: DEFAULT_FINALIZE,
+		}
+	}
+}
+
+/// HeartbeatStream is combining transaction pool notification stream and generate a new block
+///   when a certain time has passed without any transactions.
```
Comment (kianenigma) [inline client/consensus/manual-seal/src/heartbeat_stream.rs:55]:
```suggestion
/// when a certain time has passed without any transactions.
```

File: client/consensus/manual-seal/src/heartbeat_stream.rs [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+pub struct HeartbeatStream<Hash> {
+	/// The transaction pool notification stream
+	pool_stream: Box<dyn Stream<Item = EngineCommand<Hash>> + Unpin + Send>,
+	/// Future to control when the next block should be generated
+	delay_future: Delay,
+	/// To remember when the last block is generated
+	last_blocktime: Option<Instant>,
+	/// Heartbeat options
+	opts: HeartbeatOptions,
```
Comment (kianenigma) [inline client/consensus/manual-seal/src/heartbeat_stream.rs:64]:
```suggestion
	options: HeartbeatOptions,
```

(opinionated)
I'd keep all variable names un-abbreviated except special 1 letter cases.

File: client/consensus/manual-seal/src/heartbeat_stream.rs [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+			pool_stream,
+			delay_future: tokio::time::delay_for(Duration::from_secs(opts.max_blocktime)),
+			last_blocktime: None,
+			opts
+		}
+	}
+}
+
+impl<Hash> Stream for HeartbeatStream<Hash> {
+
```
Comment (kianenigma) [inline client/consensus/manual-seal/src/heartbeat_stream.rs:85]:
```suggestion
```

File: client/consensus/manual-seal/src/lib.rs (hunk 9 above)
Comment (kianenigma) [reply client/consensus/manual-seal/src/lib.rs:22]: FYI pubic items can be unused in a library.

**Review by kianenigma — COMMENTED** [review]
So the `min_block_time` is raising a question for me. 

With instant seal, it makes sense. Basically it makes it almost-instant-seal.

But, if I run manual seal, and receive some transactions immediately, and send the RPC command immediately, all before reaching the new `min_block_time`, will it still be delayed?

File: client/consensus/manual-seal/src/heartbeat_stream.rs [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+use std::{pin::Pin};
+use futures::{
+	prelude::*,
+	task::{Context, Poll},
+};
+use tokio::time::{Duration, Instant, Delay};
+
+use crate::rpc::{EngineCommand};
+
+/// Constants for generating default heartbeat options
```
Comment (kianenigma) [inline client/consensus/manual-seal/src/heartbeat_stream.rs:29]:
```suggestion
// Constants for generating default heartbeat options
```

This is being applied to the first one and not the rest if `///`. Besides since it is not public rustdoc `///` is not needed and `//` will suffice.

File: client/consensus/manual-seal/src/seal_new_block.rs [hunk 16]
```diff
@@ -1,18 +1,20 @@
-// Copyright 2019 Parity Technologies (UK) Ltd.
 // This file is part of Substrate.
 
-// Substrate is free software: you can redistribute it and/or modify
+// Copyright (C) 2020 Parity Technologies (UK) Ltd.
```
Comment (bkchr) [inline client/consensus/manual-seal/src/seal_new_block.rs:3]:
```suggestion
// Copyright (C) 2019 Parity Technologies (UK) Ltd.
```

File: client/consensus/manual-seal/Cargo.toml [hunk 17]
```diff
@@ -20,7 +21,7 @@ jsonrpc-derive = "14.2.1"
 log = "0.4.8"
 parking_lot = "0.10.0"
 serde = { version = "1.0", features=["derive"] }
-assert_matches = "1.3.0"
+tokio = { version = "0.2", features=["full"] }
```
Comment (bkchr) [inline client/consensus/manual-seal/Cargo.toml:24]: PLease revert this change.

We use `futures-timer` for this.

File: client/consensus/manual-seal/src/heartbeat_stream.rs [hunk 18]
```diff
@@ -0,0 +1,133 @@
...
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! HeartbeatStream type and HeartbeatOptions used in instant seal
+use std::{pin::Pin};
+use futures::{
+	prelude::*,
+	task::{Context, Poll},
+};
+use tokio::time::{Duration, Instant, Delay};
```
Comment (bkchr) [inline client/consensus/manual-seal/src/heartbeat_stream.rs:25]: Please use from futures-timer.

File: client/consensus/manual-seal/src/heartbeat_stream.rs [hunk 19]
```diff
@@ -0,0 +1,133 @@
...
+
+use crate::rpc::{EngineCommand};
```
Comment (bkchr) [inline client/consensus/manual-seal/src/heartbeat_stream.rs:27]:
```suggestion
use crate::rpc::EngineCommand;
```

File: client/consensus/manual-seal/src/heartbeat_stream.rs [hunk 20]
```diff
@@ -0,0 +1,133 @@
...
+
+/// Constants for generating default heartbeat options
+const DEFAULT_MAX_BLOCKTIME: u64 = 30;
```
Comment (bkchr) [inline client/consensus/manual-seal/src/heartbeat_stream.rs:30]: Do we really need them as constants?

File: client/consensus/manual-seal/src/heartbeat_stream.rs [hunk 21]
```diff
@@ -0,0 +1,133 @@
...
+	/// To remember when the last block is generated
+	last_blocktime: Option<Instant>,
+	/// Heartbeat options
+	opts: HeartbeatOptions,
+}
+
+impl<Hash> HeartbeatStream<Hash> {
+	pub fn new(
+		pool_stream: Box<dyn Stream<Item = EngineCommand<Hash>> + Unpin + Send>,
+		opts: HeartbeatOptions
```
Comment (bkchr) [inline client/consensus/manual-seal/src/heartbeat_stream.rs:70]:
```suggestion
		opts: HeartbeatOptions,
```

File: client/consensus/manual-seal/src/heartbeat_stream.rs [hunk 22]
```diff
@@ -0,0 +1,133 @@
...
+	) -> Self {
+		if opts.min_blocktime > opts.max_blocktime {
+			panic!("Heartbeat options `min_blocktime` value must not be larger than `max_blocktime` value.");
```
Comment (bkchr) [inline client/consensus/manual-seal/src/heartbeat_stream.rs:73]: Nope! No panics!

Return an `Error`.

File: client/consensus/manual-seal/src/heartbeat_stream.rs [hunk 23]
```diff
@@ -0,0 +1,133 @@
...
+		}
+		Self {
+			pool_stream,
+			delay_future: tokio::time::delay_for(Duration::from_secs(opts.max_blocktime)),
+			last_blocktime: None,
+			opts
```
Comment (bkchr) [inline client/consensus/manual-seal/src/heartbeat_stream.rs:79]:
```suggestion
			opts,
```

File: client/consensus/manual-seal/src/heartbeat_stream.rs [hunk 24]
```diff
@@ -0,0 +1,133 @@
...
+				if let Some(last_blocktime) = hbs.last_blocktime {
+					// If the last block happened less than the min_blocktime time, we want to wait till
+					//   `min_blocktime` has lapsed.
+					let passed_blocktime = Instant::now() - last_blocktime;
+					if passed_blocktime < Duration::from_secs(hbs.opts.min_blocktime) {
+						// We set `delay_future` here so it will wake up when the min_blocktime since the last
+						//   block created is passed.
+						hbs.delay_future = tokio::time::delay_for(Duration::from_secs(hbs.opts.min_blocktime)
+							- passed_blocktime);
+						return Poll::Pending;
```
Comment (bkchr) [inline client/consensus/manual-seal/src/heartbeat_stream.rs:101]: This will not work. When you create a new future, you need to poll it, so it is registered to wake up the future.

File: client/consensus/manual-seal/src/heartbeat_stream.rs [hunk 25]
```diff
@@ -0,0 +1,133 @@
...
+
+	type Item = EngineCommand<Hash>;
+
+	fn poll_next(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
+		let mut hbs = self.get_mut();
+		match hbs.pool_stream.poll_next_unpin(cx) {
+			Poll::Ready(Some(ec)) => {
+				if let Some(last_blocktime) = hbs.last_blocktime {
+					// If the last block happened less than the min_blocktime time, we want to wait till
+					//   `min_blocktime` has lapsed.
```
Comment (bkchr) [inline client/consensus/manual-seal/src/heartbeat_stream.rs:94]:
```suggestion
					//  `min_blocktime` has lapsed.
```

File: client/consensus/manual-seal/src/heartbeat_stream.rs [hunk 26]
```diff
@@ -0,0 +1,133 @@
...
+			opts
+		}
+	}
+}
+
+impl<Hash> Stream for HeartbeatStream<Hash> {
+
+	type Item = EngineCommand<Hash>;
+
+	fn poll_next(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
```
Comment (bkchr) [inline client/consensus/manual-seal/src/heartbeat_stream.rs:88]: The logic in here is confusing.

When I understood it correctly, you always only want to generate a block at a rate of `min_blocktime` and at maximum `max_blocktime` between blocks.

So what I would propose the following:

1. You always poll `delay_future` first. 
2. If the `delay_future` fired, check if the `pool_stream` had any events, if there was an event build a new block.
3. If there was no event, check if the `maximum_block` time passed and build a block if so.

`delay_future` should be changed to an interval to make it easier and you should also introduce an interval for the maximum block time.

File: client/consensus/manual-seal/src/lib.rs [hunk 27]
```diff
@@ -19,7 +19,13 @@
...
+
+use std::{sync::Arc, marker::{PhantomData, Send}};
```
Comment (bkchr) [inline client/consensus/manual-seal/src/lib.rs:24]:
```suggestion
use std::{sync::Arc, marker::PhantomData};
```

`Send` is in prelude.

File: client/consensus/manual-seal/src/lib.rs [hunk 28]
```diff
@@ -171,286 +184,29 @@ pub async fn run_instant_seal<B, CB, E, C, A, SC, T>(
...
 			EngineCommand::SealNewBlock {
 				create_empty: false,
-				finalize: false,
+				finalize,
 				parent_hash: None,
 				sender: None,
 			}
 		});
 
+	let stream: Box<dyn Stream<Item=EngineCommand<<B as BlockT>::Hash>> + Unpin + Send> =
```
Comment (bkchr) [inline client/consensus/manual-seal/src/lib.rs:196]: Does it really require the full type here?

**Comment by JoshOrndorff** [comment]
I think these changes should be made to manual seal, so that they affect both the manual seal, and instant seal which uses manual seal internally. So @kianenigma, yes I imagine the minimum block time affecting the manual seal the exact same way. Any particular node hat doesn't want that behavior can turn it off by setting the delay to zero.

# Timing Daigrams
I made some timing diagrams that I think will help us all understand the way this is supposed to work, which should make it easier to review the actual code. The horizontal axis is time. The pink marks above the axis indicate a "trigger" which means an RPC call in the case of manual seal or a transaction in the case of instant seal. The white marks below the axis indicate when a block is authored.

## Before this PR

Simple, a block is authored with every trigger. If there are no triggers, there are no blocks. If there are a bunch of really fast triggers, there are a bunch of really fast blocks (I was able to get over 1000 blocks/sec). This is how it should still work when both the min and max time are disabled.
image

## Max Block Time (aka "heartbeat blocks")

If the specified time elapses with no triggers, a heartbeat block is authored anyway. This helps users and offchain tools know the chain is still live. In the figure the max block time is five time units.

image

## Min Block Time

Introduces a "cool down" period after each block is authored in which no other blocks can be authored. If any triggers arrive during the cool down period, they are suppressed, and a single block is authored after the cooldown period. This single block ensures that at least one block is authored after each trigger, which is particularly useful in instant seal where it will clear any transactions that have arrived in the pool. The cooldown periods are shown just below the axis, and are 1 time unit long.

image

# How many timers?

> you should also introduce an interval for the maximum block time.

@bkchr, I initially also thought there should be two times. One for max block time, one for min block time. Jimmy changed my mind when he pointed out that there will always be exactly one block scheduled at any point in time. If a trigger arrives and a block is authored, then you schedule a block at the maximum time, replacing any previously scheduled block. If a trigger arrives and no block is authored because the min time has not yet elapsed, then you schedule a block at the minimum block time from the last authored block, replacing any previously scheduled block.

Now that I've written this out, I think I actually think I like the two-timer design better.

File: client/consensus/manual-seal/src/heartbeat_stream.rs (hunk 20 above)
Comment (JoshOrndorff) [reply client/consensus/manual-seal/src/heartbeat_stream.rs:30]: They won't be used anywhere else outside of this file. Should they just be variables? Or just hard-coded into the `new` function?

Comment (JoshOrndorff) [reply client/consensus/manual-seal/src/heartbeat_stream.rs]: I've moved them to the `default` function.

File: client/consensus/manual-seal/src/lib.rs (hunk 9 above)
Comment (JoshOrndorff) [reply client/consensus/manual-seal/src/lib.rs]: I removed this line. It was actually hiding a valuable warning that `rpc::ManualSeal` was no longer public which is also fixed.

File: client/consensus/manual-seal/src/lib.rs (hunk 28 above)
Comment (JoshOrndorff) [reply client/consensus/manual-seal/src/lib.rs]: No. I've reduced it to `Box<dyn Stream<Item = _> + Unpin>`. I don't think it can be reduced further. This line may become obsolete after we implement your other changes anyway.

**Comment by JoshOrndorff** [comment]
I've changed my mind. I no longer think the min_block_time should be applied to manual seal. When using manual seal you are able to manually specify which block to use as the parent. Enforcing a minimum block time means suppressing rpc calls that come too close together and replacing them by a single authoring event later. If the RPC calls specify different parents, then we would somehow have to choose which parent to author on. That sounds like a mess.

So my new opinion is that **heartbeat blocks should be available to both instant- and manual- seal, but min-block-time should only be available to instant-seal**.

**Comment by jimmychu0807** [comment]
My thought is keeping block authoring and finalizing in manual-seal totally manual, aka sending a RPC instruction to author block, and another RPC instruction to finalize block. So it is as dumb as it can be. Then we do all the smart logic in the instant-seal. So I prefer to keep the `max_blocktime` and `cooldown_blocktime` (formerly called `min_blocktime`, but I found `cooldown_blocktime` convey its meaning better) in instant-seal section.

File: client/consensus/manual-seal/src/heartbeat_stream.rs (hunk 13 above)
Comment (jimmychu0807) [reply client/consensus/manual-seal/src/heartbeat_stream.rs]: updated

File: client/consensus/manual-seal/src/heartbeat_stream.rs (hunk 22 above)
Comment (jimmychu0807) [reply client/consensus/manual-seal/src/heartbeat_stream.rs]: I returned the error upstream. But the upstream caller is an async function of `run_instant_seal`. Can I use an `expect` there, or else how do I return a `Result` further up in an async function?

File: client/consensus/manual-seal/src/finalize_block.rs (hunk 10 above)
Comment (jimmychu0807) [reply client/consensus/manual-seal/src/finalize_block.rs:1]: I find this seems to be the latest way of license. It also updating the year to 2020

File: client/consensus/manual-seal/src/heartbeat_stream.rs (hunk 24 above)
Comment (jimmychu0807) [reply client/consensus/manual-seal/src/heartbeat_stream.rs]: The latest commit should fix this issue.

**Comment by jimmychu0807** [comment]
@bkchr @seunlanlege 

1. I have updated the `heartbeat` period and `cooldown` period code. Any of them can be `None`. If both are `None`, then it is just the normal instant-seal.

2. I have fixed the delay future that is not being polled. Tested it.

3. I am updating the license mainly for the year to 2020. The text is mostly copied from `frame/benchmark/src`. Let me know if you have strong opinion against them.

4. I commented two lines in the `tests.rs` (mentioned in the code) because it didn't pass when I run it. Seun, does it pass when you run? Anyway, I think it can be fixed in another PR.

Please review them and let me know if the PR is good. Thank you! 🙏

File: client/consensus/manual-seal/src/tests.rs [hunk 29]
```diff
@@ -0,0 +1,429 @@
...
+	assert!(sink.send(EngineCommand::SealNewBlock {
+		parent_hash: Some(created_block.hash),
+		sender: Some(tx2),
+		create_empty: false,
+		finalize: false,
+	}).await.is_ok());
+
+	// Ensure the forked block is in the db
+	// note: The following line fails the test
+	// let imported = rx2.await.unwrap().unwrap();
```
Comment (jimmychu0807) [inline client/consensus/manual-seal/src/tests.rs:311]: @seunlanlege This is the line that fail when I run the test. Does it pass for you?

Comment (JoshOrndorff) [reply client/consensus/manual-seal/src/tests.rs:311]: When I was working on #7070 I noticed that one of the already-existing tests was failing. Is that the case here Jimmy? Did this test fail before you made changes?

Comment (jimmychu0807) [reply client/consensus/manual-seal/src/tests.rs:311]: This is the test that fail 4/5 times approximately. Sometimes it pass...

**Comment by gnunicorn** [comment]
@jimmychu0807 please merge master

**Comment by jimmychu0807** [comment]
will do @gnunicorn

**Comment by jimmychu0807** [comment]
I think this PR is good for merge after getting the latest master @gnunicorn

@bkchr you have previously commented, want to see if you still have any concern.
@seunlanlege @andresilva  Seun, you are the original author of manual-seal and Andre you oversee consensus, let me know if you have any comment on the work.

Thank you for all your time 🙏

File: client/consensus/manual-seal/src/heartbeat_stream.rs [hunk 30]
```diff
@@ -0,0 +1,149 @@
...
+		self.delay_for = self.options.heartbeat.and_then(|hb| Some(Delay::new(hb)));
+	}
+}
+
+impl<Hash> Stream for HeartbeatStream<Hash> {
+	type Item = EngineCommand<Hash>;
+
+	fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
+		match self.command_stream.poll_next_unpin(cx) {
+
```
Comment (bkchr) [inline client/consensus/manual-seal/src/heartbeat_stream.rs:87]:
```suggestion
```

File: client/consensus/manual-seal/src/heartbeat_stream.rs [hunk 31]
```diff
@@ -0,0 +1,149 @@
...
+								sender: None,
+							}));
+						}
+
+						Poll::Pending
+					},
+					None => Poll::Pending,
+				}
+			}, // End of `Poll::Pending`
+
```
Comment (bkchr) [inline client/consensus/manual-seal/src/heartbeat_stream.rs:144]:
```suggestion
			},
```

File: client/consensus/manual-seal/src/heartbeat_stream.rs [hunk 32]
```diff
@@ -0,0 +1,149 @@
...
+	fn create_block_now_and_reset_delay(&mut self) {
+		self.last_blocktime = Some(Instant::now());
+		self.delay_for = self.options.heartbeat.and_then(|hb| Some(Delay::new(hb)));
+	}
+}
+
+impl<Hash> Stream for HeartbeatStream<Hash> {
+	type Item = EngineCommand<Hash>;
+
+	fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
```
Comment (bkchr) [inline client/consensus/manual-seal/src/heartbeat_stream.rs:85]: The logic in this function is too complicated in here.

There are also some questions that arise, what happens when we drop such an engine command? Like we should finalize a block and now it is just dropped?

And why didn't you followed my earlier proposal for this? That would not result in dropping any events.

**Comment by bkchr** [comment]
The logic of the heartbeat stream is still too complicated and it drops events. I assume that we should not drop events, because they could be rather important?

File: client/consensus/manual-seal/src/heartbeat_stream.rs [hunk 33]
```diff
@@ -0,0 +1,149 @@
...
+			Poll::Ready(Some(ec)) => {
+				// An engine command comes in, meaning a transaction in `tx_pool` now
+				let cooldown = self.options.cooldown;
+				match (self.last_blocktime, cooldown) {
+					(Some(last_blocktime), Some(cooldown)) => {
+						// Illustration on variables used below:
+						//      |------time_passed-------+------wait_further------|
+						//   last blocktime           current                  cooldown
+						//
+						let time_passed = Instant::now() - last_blocktime;
```
[the remark this answers is not in the document]
Comment (jimmychu0807) [reply client/consensus/manual-seal/src/heartbeat_stream.rs]: Updated to use `saturating_duration_since` and `checked_sub`

**Comment by jimmychu0807** [comment]
IMG_2977

@bkchr 
- Just want to confirm the understanding with you. The current implementation handles the above 3 cases ( (2), (3), and (4) in the diagram) - this is also conveying the same idea as Joshy's diagram way above. With your suggested implementation, does it handle (3) in a simpler manner? (I can't seem to find a way to do it without similar complexity as current one).

- Also on your raised concern that events are dropped: events are dropped only in case (2), when consecutive ECs come in a burst. But it doesn't matter, because new block has already been scheduled to be sealed after the cooldown. ECs only have two events only, one to seal unfinalized block, another to seal finalized block.

File: client/consensus/manual-seal/src/heartbeat_stream.rs [hunk 34]
```diff
@@ -0,0 +1,149 @@
...
+use crate::rpc::EngineCommand;
+
+/// Options to manage the behavior of the heartbeat stream
+struct Options {
+	// The amount of time passed that a new empty block will be generated when no transactions
+	// in the tx pool.
+	heartbeat: Option<Duration>,
+	// The cooldown duration after a block has been authored.
+	cooldown: Option<Duration>,
+	// Whether the created block is finalized of not.
```
Comment (bkchr) [inline client/consensus/manual-seal/src/heartbeat_stream.rs:37]:
```suggestion
	// Whether the created block is finalized.
```

**Comment by bkchr** [comment]
>     * Just want to confirm the understanding with you. The current implementation handles the above 3 cases ( (2), (3), and (4) in the diagram) - this is also conveying the same idea as [Joshy's diagram way above](https://github.com/paritytech/substrate/pull/6951#issuecomment-689274697). With your suggested implementation, does it handle (3) in a simpler manner? (I can't seem to find a way to do it without similar complexity as current one).

I still think that the implementation of the function could be made simpler, by simpler I mean better to read. Currently it is a huge match which requires some thinking to understand what is going on.

> 
>     * Also on your raised concern that events are dropped: events are dropped only in case (2), when consecutive ECs come in a burst. But it doesn't matter, because new block has already been scheduled to be sealed after the cooldown. ECs only have two events only, one to seal block, another to finalize block.

We still would drop the finalize block event when it comes in a cooldown, which would be a weird behavior. If the user sends this event and we just drop it.

**Comment by JoshOrndorff** [comment]
> We still would drop the finalize block event when it comes in a cooldown, which would be a weird behavior. If the user sends this event and we just drop it.

This would be addressed by something like https://github.com/paritytech/substrate/pull/7070

**Comment by gnunicorn** [comment]
Closing due to it being stale for quite a while.
